### PR TITLE
[IMP] stock: readd include exhausted product option

### PR DESCRIPTION
**Pull request by ryv-odoo**
In the inventory adjustment form, readd (remove in 13.0)
the option to include exhausted product.

[REM] stock: remove dead code

In views of stock inventory line,
the view with the id "stock_inventory_line_tree" was not used.
Then remove it, and rename the "stock_inventory_line_tree2"
into "stock_inventory_line_tree".

[REF] stock: remove and refactor quantities fetching

In order to clean the fetching of quantities for a inventory,
replace SQL query by a read_group. Also, the _computed_outdated
and _get_inventory_lines_values use the same read_group to be
consistent.

task-2115194

### CODE REVIEW

**Comment by sle-odoo** [comment]
related to https://github.com/odoo/odoo/pull/34387

File: addons/stock/models/stock_inventory.py [hunk 1]
```diff
@@ -225,10 +229,38 @@ def _get_quantities(self):
 
         return self.env['stock.quant'].read_group(domain, fields, group_by, lazy=False)
 
+    def _get_exhausted_inventory_line(self, quant_products_ids):
```
Comment (sle-odoo) [inline addons/stock/models/stock_inventory.py]: _get_exhausted_inventory_lines_vals

File: addons/stock/models/stock_inventory.py [hunk 2]
```diff
@@ -225,10 +229,38 @@ def _get_quantities(self):
...
+        '''
+        This function return inventory lines for exausted products
+        :param quant_products_ids: products ids to filter out
+        '''
```
Comment (sle-odoo) [inline addons/stock/models/stock_inventory.py]: i would prefer a docstring similar to this one

``` python
def _get_exhausted_inventory_lines_vals(self, quant_product_ids):
    """Return the values of the inventory lines to create if the user
    wants to include exhausted products. Exhausted products are products
    without quantities.

    :param quant_products_ids: the non exhausted products
    :return: a list containing the `stock.inventory.line` values to create
    :rtype: list
    """
```

File: addons/stock/models/stock_inventory.py [hunk 3]
```diff
@@ -225,10 +229,38 @@ def _get_quantities(self):
...
+        exhausted_domain = [('type', '=', 'product')]
+        if self.product_ids:
+            exhausted_domain = expression.AND([exhausted_domain, [('id', 'in', self.product_ids.ids)]])
+        if quant_products_ids:
+            exhausted_domain = expression.AND([exhausted_domain, [('id', 'not in', quant_products_ids)]])
+        exhausted_product_ids = [p['id'] for p in self.env['product.product'].search_read(exhausted_domain, ['id'])]
+        if self.location_ids:
+            location_ids = self.location_ids.ids
+        else:
+            location_ids = self.env['stock.warehouse'].search([]).mapped('lot_stock_id').ids
```
Comment (sle-odoo) [inline addons/stock/models/stock_inventory.py]: self.env['stock.warehouse'].search([]).lot_stock_id.ids

File: addons/stock/models/stock_inventory.py (hunk 2 above)
Comment (sle-odoo) [reply addons/stock/models/stock_inventory.py]: you could also write one for _get_inventory_lines_values :)

File: addons/stock/models/stock_inventory.py [hunk 4]
```diff
@@ -241,8 +273,12 @@ def _get_inventory_lines_values(self):
                 'package_id': quants_group['package_id'] and quants_group['package_id'][0] or False,
             }
             if quants_group['product_id']:
+                quant_products_ids.append(quants_group['product_id'][0])
                 line_values['product_uom_id'] = self.env['product.product'].browse(quants_group['product_id'][0]).uom_id.id
             vals.append(line_values)
+        if self.exhausted:
+            exhausted_lines = self._get_exhausted_inventory_line(quant_products_ids)
```
Comment (sle-odoo) [inline addons/stock/models/stock_inventory.py]: vals += self._get_exhausted_inventory_line(quant_products_ids) should work?

File: addons/stock/models/stock_inventory.py [hunk 5]
```diff
@@ -325,18 +318,14 @@ def _compute_difference(self):
...
-            quant['owner_id'] and quant['owner_id'][0]): quant['quantity']
-            for quant in grouped_quants
+        quants_groups = self.inventory_id._get_quantities()
+        quants = {(
+            quants_group['product_id'] and quants_group['product_id'][0] or False,
+            quants_group['location_id'] and quants_group['location_id'][0] or False,
+            quants_group['lot_id'] and quants_group['lot_id'][0] or False,
+            quants_group['package_id'] and quants_group['package_id'][0] or False,
+            quants_group['owner_id'] and quants_group['owner_id'][0] or False):
+            quants_group['quantity'] for quants_group in quants_groups
```
Comment (sle-odoo) [inline addons/stock/models/stock_inventory.py]: maybe _get_quantities could return this struct ? (quants)

and in _get_inventory_lines_values you could do
``` python
        for (product_id, location_id, lot_id, package_id, owner_id), quantity in quants:
            line_values = {
                'inventory_id': self.id,
                'product_qty': 0 if self.prefill_counted_quantity == "zero" else quantity,
                'theoretical_qty': quantity,
                'prod_lot_id': lot_id.id,
                'partner_id': owner_id.id,
                'product_id': product_id.id,
                'location_id': location_id.id,
                'package_id': package_id.id,
            }
```

File: addons/stock/models/stock_inventory.py [hunk 6]
```diff
@@ -239,16 +242,47 @@ def _get_quantities(self):
...
+        without quantities or quantity equal to 0.
+
+        :param quant_products_ids: the non exhausted products
+        :return: a list containing the `stock.inventory.line` values to create
+        :rtype: list
         """
-        Return the values of the inventory lines to create for this inventory.
+        self.ensure_one()
+        vals = []
+        exhausted_domain = [('type', '=', 'product')]
```
Comment (amoyaux) [inline addons/stock/models/stock_inventory.py]:
``` python3
[('type', '=', 'product'), ('active', '=', True)]
```

File: addons/stock/models/stock_inventory.py [hunk 7]
```diff
@@ -239,16 +242,47 @@ def _get_quantities(self):
...
+        if self.product_ids:
+            exhausted_domain = expression.AND([exhausted_domain, [('id', 'in', self.product_ids.ids)]])
+        if quant_products_ids:
+            exhausted_domain = expression.AND([exhausted_domain, [('id', 'not in', quant_products_ids)]])
```
Comment (amoyaux) [inline addons/stock/models/stock_inventory.py]: Failing usecase:
- 1 product in shelf1
- Do an inventory adjustement for stock, shelf1 and shelf2
 Probably missing lines for shelf2 and stock

Comment (amoyaux) [reply addons/stock/models/stock_inventory.py]: Quand on ne l'a pas je pense qu'on avait dit juste une ligne pour la location la "plus" parent de chaque entrepôt (WH/Stock mais pas tous ses enfants) donc on aurait également une ligne/entrepôt. J'aurais tendance à faire pareil si tu as des produits en stock, si tu as min une ligne par entrepôt, on n'en crée pas d'autre, si tu n'en as pas dans certains entrepôts, on te crée également une ligne.

File: addons/stock/models/stock_inventory.py [hunk 8]
```diff
@@ -205,51 +205,64 @@ def action_view_related_move_lines(self):
...
+        locations_ids = [l['id'] for l in self.env['stock.location'].search_read(domain_loc, ['id'])]
 
-        vals = []
-        Product = self.env['product.product']
-        # Empty recordset of products available in stock_quants
-        quant_products = self.env['product.product']
+        domain = [('quantity', '!=', '0'), ('product_id.active', '=', True), ('location_id', 'in', locations_ids)]
 
-        # If inventory by company
         if self.company_id:
```
Comment (Whenrow) [inline addons/stock/models/stock_inventory.py]: don't think this can be falsy

File: addons/stock/models/stock_inventory.py [hunk 9]
```diff
@@ -347,8 +348,7 @@ def _compute_outdated(self):
                 line.location_id.id,
                 line.prod_lot_id.id,
                 line.package_id.id,
-                line.partner_id.id,
-                ), 0
+                line.partner_id.id), 0
```
Comment (Whenrow) [inline addons/stock/models/stock_inventory.py]: @sle-odoo did you approuve this useless diff ?

File: addons/stock/models/stock_inventory.py [hunk 10]
```diff
@@ -239,9 +242,41 @@ def _get_quantities(self):
...
+        :param non_exhausted_set: set of tuple (product_id, location_id) of non exhausted product-location
+        :return: a list containing the `stock.inventory.line` values to create
+        :rtype: list
         """
-        Return the values of the inventory lines to create for this inventory.
+        self.ensure_one()
+        if self.product_ids:
+            product_ids = self.product_ids.ids
+        else:
+            product_ids = self.env['product.product'].search_read([('type', '=', 'product'), ('active', '=', True)], ['id'])
```
Comment (Whenrow) [inline addons/stock/models/stock_inventory.py]: add company in domain

File: addons/stock/models/stock_inventory.py [hunk 11]
```diff
@@ -260,9 +295,10 @@ def _get_inventory_lines_values(self):
                 'location_id': location_id,
                 'package_id': package_id
             }
-            if product_id:
-                line_values['product_uom_id'] = self.env['product.product'].browse(product_id).uom_id.id
+            line_values['product_uom_id'] = self.env['product.product'].browse(product_id).uom_id.id
```
Comment (Whenrow) [inline addons/stock/models/stock_inventory.py]: Could you move this change in your first commit ? thanks

**Comment by sle-odoo** [comment]
robodoo r+ rebase-ff

**Comment by sle-odoo** [comment]
robodoo r-
multicompany

**Comment by sle-odoo** [comment]
robodoo r+ rebase-ff

**Pull request closed without merging**